src/ser.rs

```diff
@@ -382,8 +400,8 @@ impl<'a, 'b, 'c> ser::SerializeMap for MapSerializer<'a, 'b, 'c> {
   type Error = Error;
 
   fn serialize_key<T: ?Sized + Serialize>(&mut self, key: &T) -> Result<()> {
-    debug_assert!(self.next_key.is_none());
-    self.next_key = Some(key.serialize(Serializer::new(self.scope))?);
+    let key = key.serialize(Serializer::new(self.scope))?;
+    self.keys.push(key.try_into().unwrap());
```

Comment:
It's not clear to me from context: is the .try_into() really infallible?

Reply:
That's true, we're guaranteed to have strings when encoding structs and enum variant to objects, but in this case Maps could be keyed by numbers in Rust or even arbitrary structs (implementing sufficient traits).

Ultimately this code path is mainly intended for `serde_json`'s `Map` which is guaranteed to have `String` keys, otherwise given the ambiguity I would always recommend returning maps as a vec of entries.

We also can't broadly map these to `v8::Map`s since that would break expectations for `serde_json::Map/Value` and for non-primitive keys in JS maps have no concept of value-equality so all keying is done via referential equality.

So to conclude, this path is intended for `serde_json::Map` which has String keys, everything else should be returned as "entries" tuple vecs IMO

Reply:
Added some code so it fails with an explanatory message instead of a `panic`. I still broadly recommend users to avoid serializing Maps for the reasons above